File: 0036-valid-sudoku/0036-valid-sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        Map<Integer, Set<Character>> rowMap = new HashMap();
        Map<Integer, Set<Character>> colMap = new HashMap();
        
        for(int i=0; i<board.length; i++) {
            for(int j=0; j<board[0].length; j++) {
                if(board[i][j] != '.') {
                    Set<Character> rowSet = rowMap.getOrDefault(i, new HashSet<Character>());
                    if(rowSet.contains(board[i][j])) {
                        // System.out.println("-----");
                        return false;
                    }
                    rowSet.add(board[i][j]);
                    rowMap.put(i, rowSet);
                    // System.out.println("-----: " + i + " : " + rowSet.toString());
                }
            }
        }
        
        for(int i=0; i<board.length; i++) {
            for(int j=0; j<board[0].length; j++) {
                if(board[j][i] != '.') {
                    Set<Character> colSet = colMap.getOrDefault(i, new HashSet<Character>());
                    if(colSet.contains(board[j][i])) {
                        // System.out.println("-----11: ");
                        return false;
                    }
                    colSet.add(board[j][i]);
                    colMap.put(i, colSet);
                    // System.out.println("-----11: " + i + " : " + colSet.toString());
                }
            }
        }
        
        //check in sub-grid;
        int xOffset = 0;
        int yOffset = 0;
        for(int k=1; k<=9; k++) {
            // System.out.println("xxxxx: " + xOffset);
            // System.out.println("yyyyy: " + yOffset);
            Set<Character> subGrid = new HashSet();
            for(int i=xOffset; i<xOffset+3; i++) {
                for(int j=yOffset; j<yOffset+3; j++) {
                    if(board[i][j] != '.') {
                        if(subGrid.contains(board[i][j])) {
                            // System.out.println("-----22: " + subGrid.toString() + " : " + i + " : " + j);
                            return false;
                        }
                        subGrid.add(board[i][j]);
                    }
                }
            }
            if(yOffset+3 >= 8) {
                xOffset += 3;
                yOffset = 0;
            } else {
                // xOffset = 0;
                yOffset += 3;
            }
        }
        
        return true;
    }
}

File: 0036-valid-sudoku/0036-valid-sudoku.js
/**
 * @param {character[][]} board
 * @return {boolean}
 */
var isValidSudoku = function(board) {
    //row wise
    for(let i=0; i<9; i++) {
        const set = new Set()
        for(let j=0; j<9; j++) {
            if(board[i][j] === ".") continue
            if(set.has(board[i][j])) return false
            set.add(board[i][j])
        }
    }
    //columns
    for(let i=0; i<9; i++) {
        const set = new Set()
        for(let j=0; j<9; j++) {
            if(board[j][i] === ".") continue
            if(set.has(board[j][i])) return false
            set.add(board[j][i])
        }
    }


    let m=0, n=0, grid = 0
    while(grid < 9) {
        const set = new Set()
        if(m === 9) {
            n += 3
            m = 0
        }
        for(let i=m; i<m+3; i++) {
            for(let j=n; j<n+3; j++) {
                if(board[i][j] === ".") continue
                if(set.has(board[i][j])) return false
                set.add(board[i][j])
            }
        }
        m+=3
        grid++
    }
    return true
};